### PR TITLE
added all swerve drive code

### PR DESCRIPTION
**Pull request by mirapanigrahy**

### CODE REVIEW

**Comment by mirapanigrahy** [comment]
@AG6GR could you approve this merge

File: src/main/java/frc/robot/TeleopInput.java [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 // WPILib Imports
 import edu.wpi.first.wpilibj.Joystick;
+import edu.wpi.first.wpilibj.XboxController;
```
Comment (AG6GR) [inline src/main/java/frc/robot/TeleopInput.java:5]: Are we using the xbox controller for competition?

File: src/main/java/frc/robot/systems/DriveFSMSystem.java [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+				directionSlewRate = DriveConstants.INSTANTANEOUS_SLEW_RATE;
+				//some high number that means the slewrate is effectively instantaneous
+			}
+
+			double currentTime = WPIUtilJNI.now() * DriveConstants.TIME_CONSTANT;
+			double elapsedTime = currentTime - prevTime;
+			double angleDif = SwerveUtils.angleDifference(inputTranslationDir,
+				currentTranslationDir);
+
+			if (angleDif < DriveConstants.ANGLE_MULTIPLIER_1 * Math.PI) {
```
Comment (AG6GR) [inline src/main/java/frc/robot/systems/DriveFSMSystem.java:271]: What are the units of angle here?

File: src/main/java/frc/robot/systems/DriveFSMSystem.java [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+		var swerveModuleStates = DriveConstants.DRIVE_KINEMATICS.toSwerveModuleStates(
+			fieldRelative
+				? ChassisSpeeds.fromFieldRelativeSpeeds(xSpeedDelivered, ySpeedDelivered,
+					rotDelivered, Rotation2d.fromDegrees(-gyro.getAngle()))
+				: new ChassisSpeeds(xSpeedDelivered, ySpeedDelivered, rotDelivered));
+		SwerveDriveKinematics.desaturateWheelSpeeds(
+			swerveModuleStates, DriveConstants.MAX_SPEED_METERS_PER_SECOND);
+
+		int idx = 0;
+		frontLeft.setDesiredState(swerveModuleStates[idx]);
```
Comment (AG6GR) [inline src/main/java/frc/robot/systems/DriveFSMSystem.java:315]: Use actual indices 0,1,2 etc. No need to create an extra variable.

File: src/main/java/frc/robot/systems/DriveFSMSystem.java [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+		rearLeft.setDesiredState(swerveModuleStates[++idx]);
+		rearRight.setDesiredState(swerveModuleStates[++idx]);
+	}
+
+	/**
+	 * Balances gyro.
+	 */
+	public void balance() {
+		double power;
+		if (Math.abs(gyro.getRoll()) < 2 && Math.abs(gyro.getRoll()) > (-1 * 2)) {
```
Comment (AG6GR) [inline src/main/java/frc/robot/systems/DriveFSMSystem.java:326]: Simplify this expression. Absolute value is always positive.

File: src/main/java/frc/robot/systems/DriveFSMSystem.java [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+			power = 0;
+		} else if (gyro.getRoll() > 0) {
+			power = Math.abs(gyro.getRoll())
+				/ DriveConstants.BALENCE_SPEED_INVERSE_PROPORTION;
+		} else if (gyro.getRoll() < 0) {
+			power = -Math.abs(gyro.getRoll())
+				/ DriveConstants.BALENCE_SPEED_INVERSE_PROPORTION;
+		} else {
```
Comment (AG6GR) [inline src/main/java/frc/robot/systems/DriveFSMSystem.java:334]: These last three cases are exactly the same. If val < 0, then a + val is the same as a - abs(value).

File: src/main/java/frc/robot/systems/DriveFSMSystem.java [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+	public double getHeading() {
+		return Rotation2d.fromDegrees(-gyro.getAngle()).getDegrees();
+	}
+
+	/**
+	 * Returns the turn rate of the robot.
+	 *
+	 * @return The turn rate of the robot, in degrees per second
+	 */
+	public double getTurnRate() {
```
Comment (AG6GR) [inline src/main/java/frc/robot/systems/DriveFSMSystem.java:424]: Is this ever used?

File: src/main/java/frc/robot/systems/MAXSwerveModule.java
Comment (AG6GR) [inline src/main/java/frc/robot/systems/MAXSwerveModule.java:1]: Is this file copied from the REV repo? If so cite the source with a link to the original source code.

File: vendordeps/WPILibNewCommands.json
Comment (AG6GR) [inline vendordeps/WPILibNewCommands.json:1]: We should not be including WPILib commands, please remove this dependency.

**Review by AG6GR — COMMENTED** [review]
Please make sure to cite any code you did not write yourself. I recommend adding a link to the original source repo below the original copyright comment and mentioning it in the README file.

File: src/main/java/frc/robot/TeleopInput.java (hunk 1 above)
Comment (mirapanigrahy) [reply src/main/java/frc/robot/TeleopInput.java:5]: yes

File: src/main/java/frc/robot/systems/DriveFSMSystem.java (hunk 2 above)
Comment (mirapanigrahy) [reply src/main/java/frc/robot/systems/DriveFSMSystem.java:271]: radians

**Pull request closed without merging**